### PR TITLE
Add direct defund objective

### PR DESCRIPTION
**Pull request by kerzhner**
Fixes https://github.com/statechannels/go-nitro/issues/77.

Direct defund objective follows the patterns established by the direct fund objective.

### CODE REVIEW

File: channel/channel.go [hunk 1]
```diff
@@ -143,6 +144,18 @@ func (c Channel) LatestSupportedState() (state.State, error) {
...
+	numStates := len(c.SignedStateForTurnNum)
+	keys := make([]uint64, 0, len(c.SignedStateForTurnNum))
+	for k := range c.SignedStateForTurnNum {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
+	latestTurn := keys[numStates-1]
+	return c.SignedStateForTurnNum[latestTurn]
+}
+
```
Comment (NiloCK) [inline channel/channel.go:158]: Drive-by comment:

- this function may want test coverage
- this looks correct to me, but something like

```
latestTurn := -1
for k := range c.SignedStateForTurnNum {
	if k > greatest {
		latestTurn = k
	}
}
return c.SignedStateForTurnNum[latestTurn]
```
is maybe easier to read & is o(n) rather than n^2

Comment (kerzhner) [reply channel/channel.go]: Good call! Addressed in 17abec6..58eccbf

File: protocols/directdefund/directdefund.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	if updated.Status != protocols.Approved {
+		return updated, sideEffects, WaitingForNothing, ErrNotApproved
+	}
+
+	latestSignedState, err := updated.C.LatestSignedState()
+	if err != nil {
+		return updated, sideEffects, WaitingForNothing, errors.New("The channel must contain at least one signed state to crank the defund objective")
+	}
+
+	// Sign a final state if it has not been signed
```
Comment (NiloCK) [inline protocols/directdefund/directdefund.go:119]: Maybe something like "Finalize and sign a state if no supported, finalized state exists"?

**Review by NiloCK — COMMENTED** [review]
This is our first defund objective, and I was a little confused about logic around checking existing states and applying signatures to arrive at a final, supported state.

Does it make any sense to insist on a supported, final state from the client before constructing a ddfo (entirely remove this complexity from the ddfo, place it in scope of the running channel)?

As it stands, I'm not sure that there isn't a vulnerability in the shape of:
- alice & bob have a channel with supported, final state N
- ddfo is initialized, approved in alice's waller
- bob sends a signed state N+1 to alice
- alice's `crank` finds that the latestSignedState isn't finalized, and automatically finalizes and signs it

In any case, this is great overall. Maybe I'll hold the approve until I don't believe the above attack is possible.

**Comment by kerzhner** [comment]
> As it stands, I'm not sure that there isn't a vulnerability in the shape of:

Addressed in 80dfee3.

File: protocols/directdefund/directdefund.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	} else {
+		init.Status = protocols.Unapproved
+	}
+	init.C = c.Clone()
+
+	return init, nil
+}
+
+// Public methods on the DirectDefundingObjective
+func (o Objective) Id() protocols.ObjectiveId {
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply protocols/directdefund/directdefund.go]: As far as I can tell, on my machine, VS Code does not attach a comment to a function when there is an empty line between the comment and the function.

Does 9689b03 help?

File: protocols/directdefund/directdefund.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	return updated, nil
+}
+
+// Crank inspects the extended state and declares a list of Effects to be executed
+func (o Objective) Crank(secretKey *[]byte) (Objective, protocols.SideEffects, protocols.WaitingFor, error) {
+	updated := o.clone()
+
+	sideEffects := protocols.SideEffects{}
+
+	// Input validation
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply protocols/directdefund/directdefund.go]: Agreed that the comment no longer makes sense. Removed in 4da41df.

Is the recommendation to validate inputs?

File: protocols/directdefund/directdefund.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+	if err != nil {
+		return updated, sideEffects, WaitingForFinalization, fmt.Errorf("error finding a supported state: %w", err)
+	}
+	if !latestSupportedState.IsFinal {
+		return updated, sideEffects, WaitingForFinalization, nil
+	}
+
+	// Withdrawal of funds
+	if !updated.fullyWithdrawn() {
+		// TODO before submiting a withdrawal transaction, we should check if a withdrawal transaction has already been submitted
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply protocols/directdefund/directdefund.go]: Addressed in 0be8c01.

File: protocols/directdefund/directdefund_test.go [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+		return c, err
+	}
+
+	sss := make([]state.SignedState, 1)
+	sss[0] = ss
+	allOk := c.AddSignedStates(sss)
+	if !allOk {
+		return c, errors.New("Unable to add a state to channel")
+	}
+	c.OnChainFunding = map[common.Address]*big.Int{common.HexToAddress(`0xF5A1BB5607C9D079E46d1B3Dc33f257d937b43BD`): big.NewInt(1)}
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply protocols/directdefund/directdefund_test.go]: Good catch! Addressed in 20ae771.

File: protocols/directdefund/directdefund_test.go [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+	return c, nil
+}
+
+func newTestObjective(signByBob bool) (Objective, error) {
+	o := Objective{}
+
+	toSign := []bool{true, true}
+	if !signByBob {
+		toSign = []bool{true, false}
+	}
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply protocols/directdefund/directdefund_test.go:111]: We use `signByBob` parameter to trigger https://github.com/statechannels/go-nitro/blob/20ae7713b3afec4372492f97ee12be83eea0b5e1/protocols/directdefund/directdefund_test.go#L133. I agreed that the `newTestObjective` api is confusing and open to suggestions.

File: protocols/directdefund/directdefund_test.go [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+	// The third crank. Alice is expected to enter the terminal state of the defunding protocol.
+	updated.C.OnChainFunding = types.Funds{}
+	_, se, wf, err = updated.Crank(&alicePK)
+	if err != nil {
+		t.Error(err)
+	}
+	if wf != WaitingForNothing {
+		t.Errorf(`WaitingFor: expected %v, got %v`, WaitingForWithdraw, wf)
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply protocols/directdefund/directdefund_test.go]: Good catch! Addressed in 13658d0

File: protocols/directdefund/directdefund.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+		for _, ss := range event.SignedStates {
+			if !ss.State().IsFinal {
+				return o, errors.New("direct defund objective can only be updated with final states")
+			}
+		}
+	}
+
+	updated := o.clone()
+	updated.C.AddSignedStates(event.SignedStates)
+
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply protocols/directdefund/directdefund.go:137]: Good call! Update in 5949da4

File: protocols/directdefund/directdefund_test.go [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	// The third crank. Bob is expected to enter the terminal state of the defunding protocol.
+	o.C.OnChainFunding = types.Funds{}
+	_, se, wf, err = o.Crank(&bobPK)
+	if err != nil {
+		t.Error(err)
+	}
+	if wf != WaitingForNothing {
+		t.Errorf(`WaitingFor: expected %v, got %v`, WaitingForWithdraw, wf)
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply protocols/directdefund/directdefund_test.go:333]: Updated in 85c5561.

File: protocols/directdefund/directdefund_test.go [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	expectedSE = protocols.SideEffects{}
+
+	if diff := cmp.Diff(expectedSE, se); diff != "" {
+		t.Errorf("Side effects mismatch (-want +got):\n%s", diff)
+	}
+
+	// The third crank. Bob is expected to enter the terminal state of the defunding protocol.
+	o.C.OnChainFunding = types.Funds{}
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply protocols/directdefund/directdefund_test.go]: Yep, makes sense! Updated in 5949da4.

File: protocols/directdefund/directdefund.go (hunk 2 above)
Comment (kerzhner) [reply protocols/directdefund/directdefund.go]: Updated in 8288b16.

**Comment by kerzhner** [comment]
@NiloCK Many thanks for a thorough and thoughtful review. I believe all comments have been addressed.

File: protocols/directdefund/directdefund.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (NiloCK) [reply protocols/directdefund/directdefund.go]: The original had no empty line (lines 58 & 59). My own addition of **two** lines seems to have been a typo. In any case, that does help.

File: protocols/directdefund/directdefund.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (NiloCK) [reply protocols/directdefund/directdefund.go]: Not sure if there's a specific recommendation for this case. Where inputs might lead to a malformed or unworkable state, I guess. Doesn't seem to be the case here though.

File: protocols/directdefund/directdefund.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+var ErrChannelUpdateInProgress = errors.New("cannot defund channel with unsupported, non-final latest state")
+
+// Objective is a cache of data computed by reading from the store. It stores (potentially) infinite data
+type Objective struct {
+	Status protocols.ObjectiveStatus
+	C      *channel.Channel
+}
+
+func isUpdateInProgress(c *channel.Channel) bool {
+	return !c.LatestSignedState().HasAllSignatures() && !c.LatestSignedState().State().IsFinal
```
[the remark this answers is not in the document]
Comment (NiloCK) [reply protocols/directdefund/directdefund.go]: It is!

**Comment by NiloCK** [comment]
Looking good to me now. Nice fix-ups.

This work began before `UnmarshalJSON` and `MarshalJSON` were part of the `protocols.Objective` interface, and I notice now that they are absent, which will make this objective not pass the type-checker as an objective, so not immediately consumable by the rest of the project.

I think it's appropriate to merge in as-is, and someone (me?) can follow-up with these functions.

**Review by NiloCK — APPROVED** [review]

**Comment by kerzhner** [comment]
> Looking good to me now. Nice fix-ups.
> 
> This work began before `UnmarshalJSON` and `MarshalJSON` were part of the `protocols.Objective` interface, and I notice now that they are absent, which will make this objective not pass the type-checker as an objective, so not immediately consumable by the rest of the project.

Good catch. Added https://github.com/statechannels/go-nitro/issues/356
